keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
+
+* Observing when init containers or containers have substantially different resource reservation, causing workloads to under provision nodes
+* Visualize batch compute consumption on the cluster where completed pods can be quickly excluded
+
+The implementation would be gated behind a flag in the kube-scheduler `--pod-resource-metrics` to allow distributions to opt-in to the increased cardinality, although in practice we expect the impact in scheduler to be drowned by other O(pods) cardinality in other parts of the system.
+
+The metrics would be served by the scheduler that is holding the active scheduler lease. During a failover, no metrics would be reported. This would prevent duplicate metrics from being scraped and aligns with our decision in the scheduler to not fill caches prior to election for resource usage minimization.
+
+Custom schedulers based on the scheduler code should be able to easily reuse and report these metrics, and the configuration of the cluster instrumentation to retrieve those metrics would be left as an exercise for the integrator.
+
+
+### Add recording rules consistent with this metric to describe actual resource usage
+
+In order to make these metrics useful for calculation of actual usage, we would represent the current usage in a form that can easily be queried by applying recording rules in the standard Prometheus stack and equivalents in other systems.
```

Comment:
Where those metrics will be implemented?

Given that metrics-server is what is exposing the usage metrics for the API purpose, it should probably be it.

Reply:
Actually today we don't even need metrics-server - kubelet level is sufficient today to do all resource consumption calculations and I think is a reasonable mapping for metrics-as-loose-api that we have consistently enforced.

Reply:
metrics-server is only 1 implementation of the metrics APIs, it's in no way authoritative, and I would not want to push this responsibility onto metrics API providers.